flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java

```diff
@@ -205,17 +211,23 @@ private String getWebMonitorAddress(Configuration configuration) throws Exceptio
             Preconditions.checkArgument(pipelineJars.size() == 1, "Should only have one jar");
         }
 
-        final ClusterClientProvider<String> clusterClientProvider =
-                deployClusterInternal(
-                        KubernetesApplicationClusterEntrypoint.class.getName(),
-                        clusterSpecification,
-                        false);
+        ClusterClientProvider<String> clusterClientProvider;
+        try {
+            clusterClientProvider =
+                    deployClusterInternal(
+                            KubernetesApplicationClusterEntrypoint.class.getName(),
+                            clusterSpecification,
+                            false);
 
-        try (ClusterClient<String> clusterClient = clusterClientProvider.getClusterClient()) {
-            LOG.info(
-                    "Create flink application cluster {} successfully, JobManager Web Interface: {}",
-                    clusterId,
-                    clusterClient.getWebInterfaceURL());
+            try (ClusterClient<String> clusterClient = clusterClientProvider.getClusterClient()) {
+                LOG.info(
+                        "Create flink application cluster {} successfully, JobManager Web Interface: {}",
+                        clusterId,
+                        clusterClient.getWebInterfaceURL());
+            }
+        } catch (Exception e) {
```

Comment:
I am curious whether we could wrap the `try...catch {// clean up resources}` in a separate method. Just like following. WDYT?

```
    private <T> ClusterClientProvider<T> safelyDeployCluster(
            SupplierWithException<ClusterClientProvider<T>, Exception> supplier)
            throws ClusterDeploymentException {
        try {
            return supplier.get();
        } catch (Exception e) {
            try {
                LOG.warn(
                        "Failed to create the Kubernetes cluster \"{}\", try to clean up the residual resources.",
                        clusterId);
                client.stopAndCleanupCluster(clusterId);
            } catch (Exception ex) {
                LOG.warn(
                        "Failed to stop and clean up the Kubernetes cluster \"{}\".", clusterId, e);
            }
            throw new ClusterDeploymentException(e);
        }
    }
```

Reply:
Good suggestion 👍